src/peergos/shared/user/SharedWithCache.java

```diff
@@ -15,11 +14,40 @@
     //K,V
     // K = mapKey of absolute capability
     // V = list of usernames the file/dir is shared with
-    private Map<ByteArrayWrapper, Set<String>> sharedWithReadAccessCache = new ConcurrentHashMap<>();
-    private Map<ByteArrayWrapper, Set<String>> sharedWithWriteAccessCache = new ConcurrentHashMap<>();
+    private final Map<ByteArrayWrapper, Set<String>> sharedWithReadAccessCache = new ConcurrentHashMap<>();
+    private final Map<ByteArrayWrapper, Set<String>> sharedWithWriteAccessCache = new ConcurrentHashMap<>();
+    // These would be more space efficient in a trie
+    private final Map<Path, Set<String>> writeShares = new ConcurrentHashMap<>();
```

Comment:
What about files that are moved?

Reply:
I just tested and your are right the current solution does not work with move. I'm sure it used to - hence the key of the map was not a path.

Reply:
What is the desired behaviour?

Reply:
If i move directories around, i would find it surprising that my shares stopped working. This is especially the case if i move a parent folder to a shared folder/files. Maybe revost at a later date.

Reply:
Agreed, we should think more about the situations here and what is most intuitive. For a different PR.